academy/4-my-own-chain/external-gui.md

```diff
@@ -0,0 +1,444 @@
+
+Confirm that the app works normally by running it:
+
+```sh
+$ cd checkers-gui-tmp
+$ npm start
+```
+
+It should automatically open the browser for you and you should see a simple page. If not, open it to [`http://localhost:3000`](http://localhost:3000).
+
+### Copy the React app files
+
+Now copy into your own project all the files created for the React app. If you followed from the previous section, then:
+
+* The `tsconfig.json` file, and the `public` and `src` folders have no conflicts, that is easy.
+* `.gitignore`: [just add the content](https://github.com/cosmos/academy-checkers-ui/blob/4ea0bdb/.gitignore#L3-L21) to your own `.gitignore`.
+* As for `package.json`:
+    * [Copy-paste the difference](https://github.com/cosmos/academy-checkers-ui/blob/4ea0bdb/package.json#L57-L74) for everything except the dependencies.
+    * Remove the [`"homepage"`](https://github.com/cosmos/academy-checkers-ui/blob/f9e1375/package.json#L18) field or it will confuse React.
+
+For the package dependencies, it is better to use an exact method. Click on the expandable box below:
+
+<ExpansionPanel title="Get the exact dependencies">
+
+Get the exact dependencies versions installed by the initialization script by running in the React app folder:
+
+```sh
+$ npm list @testing-library/jest-dom @testing-library/react @testing-library/user-event @types/jest @types/react @types/react-dom react react-dom react-scripts typescript web-vitals --depth=0
+```
+
+Which prints something like:
+
+```
+├── @testing-library/jest-dom@5.16.4
+├── @testing-library/react@13.1.1
+├── @testing-library/user-event@13.5.0
+├── @types/jest@27.4.1
+├── @types/react-dom@18.0.1
+├── @types/react@18.0.5
+├── react-dom@18.0.0
+├── react-scripts@5.0.1
+├── react@18.0.0
+├── typescript@4.6.3
+└── web-vitals@2.1.4
+```
+
+From here, it is just a matter of repeating this information into a new install line:
+
+```sh
+$ npm install react-dom@18.0.0 react-scripts@5.0.1 react@18.0.0 web-vitals@2.1.4 --save-exact
+$ npm install @testing-library/jest-dom@5.16.4 @testing-library/react@13.1.1 @testing-library/user-event@13.5.0 @types/jest@27.4.1 @types/react-dom@18.0.1 @types/react@18.0.5 typescript@4.6.3 --save-exact --save-dev
+```
+
+This should update your [`package.json`](https://github.com/cosmos/academy-checkers-ui/blob/4ea0bdb/package.json#L23-L55) and lock files.
+
+</ExpansionPanel>
+
+To make sure that your editing of `package.json` does not reformat it with Prettier, add the ignore file `.prettierignore` with the following in it:
+
+``` [https://github.com/cosmos/academy-checkers-ui/blob/4ea0bdb/.prettierignore#L1]
+package.json
+```
+
+### Fix a compilation error
+
+There may be a Typescript compilation error on a CosmJs generated file, so add [`"downlevelIteration": true,`](https://github.com/cosmos/academy-checkers-ui/blob/4ea0bdb/tsconfig.json#L13) to `tsconfig.json`'s `.compilerOptions`.
+
+Confirm that you get the React simple page by running `npm start` in your own project too.
```

Comment:
"in your project too". 

Not very clear. Can we identify the project folder to run from? "Project root" or something else?

Reply:
> Can we identify the project folder to run from?

If you mean _content-wise_, I am open to suggestions. The CosmJs Checkers repo contains code for the GUI and server-side scripts and more generic code in support of both.